### PR TITLE
SI-5691 lint warning when a type parameter shadows an existing type.

### PR DESCRIPTION
**Pull request by gourlaysama**
This adds a new lint warning for when a class/method/type-member's
type parameter shadows an existing type: `-Xlint:type-parameter-shadow`.

It excludes type parameters of synthetic methods (the user can't
rename or remove those anyway), otherwise, for example, every case class
triggers the warning.

Also fixes a test that contained wrong java sources (that didn't even
compile...), discovered thanks to the warning.

---

This kind of errors shows up every now and then on the mailing-list, on
stackoverflow, etc. so maybe a warning would be useful.
I was afraid this would yield too many warnings for libraries that are
heavy on type parameters, but no: running this on scalaz and shapeless
HEAD (`v7.1.0-RC1-41-g1cc0a96` and `v2.0.0-M1-225-g78426a0` respectively)
yields 44 warnings. None of them are false positives; they usually come
from:
- scalaz loving using `A` as type parameter, even several levels deep
  of parametrized classes/methods
- or calling a type parameter that will hold a map `Map`, or similar,
  thus shadowing an existing type

### CODE REVIEW

**Comment by gourlaysama** [comment]
PLS REBUILD/pr-scala@d995632

**Comment by gourlaysama** [comment]
Is it me or is `pr-scala-integrate-ide` in a bad mood this week...?

**Comment by gourlaysama** [comment]
PLS REBUILD/pr-scala@d995632

**Comment by gkossakowski** [comment]
PLS REBUILD ALL

**Comment by gourlaysama** [comment]
PLS REBUILD ALL

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -1749,6 +1749,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
              |you want, you must write the annotation class in Java.""".stripMargin)
       }
 
+      if (settings.warnTypeParameterShadow) warnTypeParameterShadow(tparams1, clazz)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: you could push the `if` into the method, so you don't need to repeat it three times

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -2133,6 +2135,22 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         failStruct(ddef.tpt.pos, "a user-defined value class", where = "Result type")
     }
 
+    // warn about class/method/type-members' type parameters that shadow types already in scope
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I'd prefer to move this code somewhere else, `Typers` is already sooo long. I think `trait TyperDiagnostics` in the file `TypeDiagnostics.scala` might be an OK place.

**Comment by lrytz** [comment]
Nice, thanks! Looks good, modulo my minor comments.

**Comment by gourlaysama** [comment]
@lrytz updated :)

**Comment by lrytz** [comment]
LGTM!

**Pull request merged**